### PR TITLE
Download dialog

### PR DESCRIPTION
**Pull request by dsamojlenko**
# Summary | Résumé

Download dialog. Enables user to:
- Download Individual HTML files + HTML receipt (ZIP only)
- Download CSV + HTML receipt (individual or ZIP)
- Download JSON + HTML receipt (individual or ZIP)

Screen Recording 2023-11-20 at 3 18 20 PM

### CODE REVIEW

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import { useTranslation } from "react-i18next";
+import { Button } from "@components/globals";
+import { logMessage } from "@lib/logger";
+import axios from "axios";
+import { DownloadFormat } from "@lib/responseDownloadFormats/types";
+import JSZip from "jszip";
+
+export const DownloadDialog = ({
+  checkedItems,
+  isVisible,
```
Comment (timarney) [inline components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:12]: Maybe rename this so it's more clean that this shows or hides the dialog.

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  const [zip, setZip] = React.useState<boolean>(true);
+
+  useEffect(() => {
+    if (selectedFormat === DownloadFormat.HTML_ZIPPED) {
+      setZip(true);
+    }
+  }, [selectedFormat]);
+
+  if (downloadError) {
+    setDownloadError(false);
```
Comment (timarney) [inline components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:39]: This seems odd -- if there's a download error set the download error to false?

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    anchorElement.download = fileName;
+    document.body.appendChild(anchorElement);
+    anchorElement.click();
+    document.body.removeChild(anchorElement);
+    window.URL.revokeObjectURL(href);
+  };
+
+  const handleDownload = async () => {
+    const url = `/api/id/${formId}/submission/download?format=${selectedFormat}`;
+    const ids = Array.from(checkedItems.keys());
```
Comment (timarney) [inline components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:62]: Should we add some empty checks here - i.e. if you don't have checked items + if ids are empty.

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        const response = await axios({
+          url,
+          method: "POST",
+          responseType: "blob",
+          data: {
+            ids: ids.join(","),
+          },
+        });
+
+        const fileName = `records.zip`;
```
Comment (timarney) [inline components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:75]: Maybe add a date to this so it's unique

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  setIsVisible: React.Dispatch<React.SetStateAction<boolean>>;
+  downloadError: boolean;
+  setDownloadError: React.Dispatch<React.SetStateAction<boolean>>;
+  formId: string;
+  onSuccessfulDownload: () => void;
+}) => {
+  const dialogRef = useDialogRef();
+  const { t } = useTranslation("form-builder-responses");
+  const [selectedFormat, setSelectedFormat] = React.useState<DownloadFormat>();
+  const [zip, setZip] = React.useState<boolean>(true);
```
[the remark this answers is not in the document]
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx]: Renamed to zipAllFiles, setZipAllFiles

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 1 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx]: renamed to isDialogVisible, setIsDialogVisible

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 3 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:62]: You wouldn't be able to reach the dialog if you didn't have items checked. But if you somehow did, you'd just get an error back from the API.

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 2 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:39]: It's resetting the download error when the dialog pops. Would this be better in a useEffect or something? It's basically if for some reason you attempted but got a download error, if you pop the dialog again it clears the error.

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 4 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:75]: Going to work on filenames with Carolyn today

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 2 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:39]: Fixed underlying bug

File: components/form-builder/app/responses/Dialogs/DownloadDialog.tsx (hunk 4 above)
Comment (dsamojlenko) [reply components/form-builder/app/responses/Dialogs/DownloadDialog.tsx:76]: Will add in separate PR

**Pull request merged**